File: apps/web/src/net/protocol.ts
import type { Action, GameConfig } from '@ttr/engine';

export type RoomStatus = 'lobby' | 'playing' | 'ended';

export interface PlayerSnapshot {
  id: string;
  name: string;
  ready: boolean;
}

export interface CreateRoomMsg {
  type: 'createRoom';
  playerName: string;
}

export interface JoinRoomMsg {
  type: 'joinRoom';
  code: string;
  playerName: string;
  playerId?: string;
  lastSeq?: number;
}

export interface RoomStateMsg {
  type: 'roomState';
  code: string;
  hostId: string;
  players: PlayerSnapshot[];
  status: RoomStatus;
  seed?: number;
  seq: number;
  youId: string;
}

export interface SetReadyMsg {
  type: 'setReady';
  ready: boolean;
}

export interface StartGameMsg {
  type: 'startGame';
  config: Omit<GameConfig, 'seed'> & { seed?: number };
}

export interface ActionMsg {
  type: 'action';
  action: Action;
}

export interface ActionAppliedMsg {
  type: 'actionApplied';
  seq: number;
  action: Action;
}

export interface LeaveMsg {
  type: 'leave';
}

export interface ErrorMsg {
  type: 'error';
  code: string;
  message: string;
}

export interface PingMsg {
  type: 'ping';
  t: number;
}

export interface PongMsg {
  type: 'pong';
  t: number;
}

export type ClientToServer =
  | CreateRoomMsg
  | JoinRoomMsg
  | SetReadyMsg
  | StartGameMsg
  | ActionMsg
  | LeaveMsg
  | PingMsg;

export type ServerToClient = RoomStateMsg | ActionAppliedMsg | ErrorMsg | PongMsg;

File: apps/server/internal/ws/protocol.go
// Package ws defines the WebSocket message protocol and HTTP handler.
package ws

import "encoding/json"

// Message type discriminators (lowerCamelCase to match TS client).
const (
	TypeCreateRoom    = "createRoom"
	TypeJoinRoom      = "joinRoom"
	TypeRoomState     = "roomState"
	TypeSetReady      = "setReady"
	TypeStartGame     = "startGame"
	TypeAction        = "action"
	TypeActionApplied = "actionApplied"
	TypeLeave         = "leave"
	TypeError         = "error"
	TypePing          = "ping"
	TypePong          = "pong"
)

// Status values for a room.
const (
	StatusLobby   = "lobby"
	StatusPlaying = "playing"
	StatusEnded   = "ended"
)

// Envelope is the wire-level discriminator; raw payload is decoded per type.
type Envelope struct {
	Type string          `json:"type"`
	Raw  json.RawMessage `json:"-"`
}

type CreateRoomMsg struct {
	Type       string `json:"type"`
	PlayerName string `json:"playerName"`
}

type JoinRoomMsg struct {
	Type       string `json:"type"`
	Code       string `json:"code"`
	PlayerName string `json:"playerName"`
	PlayerID   string `json:"playerId,omitempty"`
	LastSeq    int    `json:"lastSeq,omitempty"`
}

type PlayerSnapshot struct {
	ID    string `json:"id"`
	Name  string `json:"name"`
	Ready bool   `json:"ready"`
}

type RoomStateMsg struct {
	Type    string           `json:"type"`
	Code    string           `json:"code"`
	HostID  string           `json:"hostId"`
	Players []PlayerSnapshot `json:"players"`
	Status  string           `json:"status"`
	Seed    *int64           `json:"seed,omitempty"`
	Seq     int              `json:"seq"`
	YouID   string           `json:"youId"`
}

type SetReadyMsg struct {
	Type  string `json:"type"`
	Ready bool   `json:"ready"`
}

type StartGameMsg struct {
	Type   string          `json:"type"`
	Config json.RawMessage `json:"config"`
}

type ActionMsg struct {
	Type   string          `json:"type"`
	Action json.RawMessage `json:"action"`
}

type ActionAppliedMsg struct {
	Type   string          `json:"type"`
	Seq    int             `json:"seq"`
	Action json.RawMessage `json:"action"`
}

type LeaveMsg struct {
	Type string `json:"type"`
}

type ErrorMsg struct {
	Type    string `json:"type"`
	Code    string `json:"code"`
	Message string `json:"message"`
}

type PingMsg struct {
	Type string `json:"type"`
	T    int64  `json:"t"`
}

type PongMsg struct {
	Type string `json:"type"`
	T    int64  `json:"t"`
}
